lib/map/security_token/service.rb

```diff
@@ -8,15 +8,15 @@ module SecurityToken
     class Service < Common::Client::Base
       configuration Configuration
 
-      def token(application:, icn:)
+      def token(application:, icn:, cache: true)
         Rails.logger.info("#{config.logging_prefix} token request", { application:, icn: })
-        response = perform(:post,
-                           config.token_path,
-                           token_params(application, icn),
-                           { 'Content-Type' => 'application/x-www-form-urlencoded' })
-        sts_token = parse_response(response, application, icn)
-        Rails.logger.info("#{config.logging_prefix} token success", { application:, icn: })
-        sts_token
+
+        if cache && (cached_token = Rails.cache.read("map_sts_token_#{application}_#{icn}"))
+          Rails.logger.info("#{config.logging_prefix} token success", { application:, icn:, cache: })
+          cached_token
+        else
+          request_token(application:, icn:)
+        end
```

Comment:
[Rails.cache.fetch](https://api.rubyonrails.org/classes/ActiveSupport/Cache/Store.html#method-i-fetch) will give all of this to you for free. Basically when you call `fetch` it will try to fetch from the cache with the key, if it hits it will return the cached value, if it misses it will execute the block then store and return the result. Also, if you pass `force: true` to fetch it will force a miss and always run the block 

```ruby
# First time foo is called, the cache will miss so it will run the block and store the result.
# Subsequent calls to foo will hit return the cached value until the cache expires.

def foo(use_cache: true)
  cache_result = :hit

  bar = Rails.cache.fetch('foobar', expires_in: 5.minutes, force: !use_cache) do
    cache_result = :missed
    'baz'
  end

  Rails.logger.info("Cache #{cache_result}")
  bar
end
```
```bash
3.2.4 :013 > foo
2024-05-21 14:25:54.812540 I [65053:12180 (irb):8] Rails -- Cache missed
 => "baz" 
3.2.4 :014 > foo
2024-05-21 14:25:59.055084 I [65053:12180 (irb):8] Rails -- Cache hit
 => "baz" 
3.2.4 :015 > foo
2024-05-21 14:26:02.923146 I [65053:12180 (irb):8] Rails -- Cache hit
 => "baz" 
3.2.4 :016 > foo(use_cache: false)
2024-05-21 14:26:18.717945 I [65053:12180 (irb):8] Rails -- Cache missed
 => "baz"
```